### PR TITLE
use buffered channel to manage items in query pool

### PR DESCRIPTION
**Pull request by adwski**
## Pull request type

Please check the type of change your PR introduces:

- [x] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Current approach has couple shortcomings:
- items can be created above configured limit
- item will not be closed during shutdown if it was only in idle slice and not added to index map.
- pool has no warm-up, new items are created upon request 

## What is the new behavior?

I propose to use buffered channel to manage query pool items as oppose to index map and idle slice. This approach makes item creation process continuous and asynchronous, which allows seamlessly warm-up pool or re-establish all connections after connectivity failure.

Internally I introduced: 
- item queue (buffered channel) - main item queue
- item tokens (buffered channel) - item creation requests
- item spawner goroutine - runs asynchronously and fills queue with new items

getItem() and putItem() are working with item queue - receiving or sending back items. New thing about getItem() is that it ensures that retrieved item is ok to use (as far IsAlive() can tell). To achieve this it spins in endless loop until in can receive item from queue. This means in case of connectivity failure or pool starvation it will block causing back pressure to caller (who should use context with deadline if it wishes to not block as well but detect this situation).

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,5 @@
+* Added warm-up and asynchronous item handling in internal/pool
```
Comment (asmyasnikov) [inline CHANGELOG.md:1]: no need to add a line for internal refactoring changes

Comment (asmyasnikov) [reply CHANGELOG.md:1]: Changelog for only user-facing changes

File: internal/pool/pool.go [hunk 2]
```diff
@@ -172,16 +197,61 @@ func New[PT Item[T], T any](
...
+		case <-p.done:
+			break spawnLoop
+		case <-p.itemTokens:
+			// got token, must create item
+			for {
+				item, err := p.createItem(ctx)
+				if err != nil {
+					select {
+					case <-ctx.Done():
+						break spawnLoop
```
Comment (asmyasnikov) [inline internal/pool/pool.go:230]: no need a label for loop - use return instead

File: internal/pool/pool.go [hunk 3]
```diff
@@ -172,16 +197,61 @@ func New[PT Item[T], T any](
...
 
-	p.idle = make([]PT, 0, p.limit)
-	p.index = make(map[PT]struct{}, p.limit)
 	p.stats = &safeStats{
 		v:        stats.Stats{Limit: p.limit},
 		onChange: p.trace.OnChange,
 	}
 
+	for i := 0; i < defaultSpawnGoroutinesNumber; i++ {
+		go p.spawnItems(ctx)
```
Comment (asmyasnikov) [inline internal/pool/pool.go:206]: context must be avalaible only on New call
After return managment to upper code - context maybe is done
For spawn goroutines - you can use xcontext.ValueOnly(ctx)

File: internal/pool/pool.go [hunk 4]
```diff
@@ -172,16 +197,61 @@ func New[PT Item[T], T any](
...
+			// got token, must create item
+			for {
+				item, err := p.createItem(ctx)
+				if err != nil {
+					select {
+					case <-ctx.Done():
+						break spawnLoop
+					case <-p.done:
+						break spawnLoop
+					case <-time.After(defaultCreateRetryDelay):
```
Comment (asmyasnikov) [inline internal/pool/pool.go:233]: time.After produce timer inside - this can be provokes the goroutines/rosurces leak
You can create ticker/timer before select - and call directly ticker.Stop() in defer

File: internal/pool/defaults.go [hunk 5]
```diff
@@ -1,6 +1,22 @@
...
 
-const DefaultLimit = 50
+import "time"
+
+const (
+	DefaultLimit = 50
+
+	// defaultCreateRetryDelay specifies amount of time that spawner will wait
+	// before retrying unsuccessful item create.
+	defaultCreateRetryDelay = 500 * time.Millisecond
```
Comment (asmyasnikov) [inline internal/pool/defaults.go:10]: Can this setting helps for exactly once case?
I think this is over-engineering without the done cases

File: internal/pool/pool.go [hunk 6]
```diff
@@ -358,25 +397,29 @@ func (p *Pool[PT, T]) putItem(ctx context.Context, item PT) (finalErr error) {
 	select {
 	case <-p.done:
 		return xerrors.WithStackTrace(errClosedPool)
+	case <-ctx.Done():
+		return xerrors.WithStackTrace(ctx.Err())
 	default:
-		if !item.IsAlive() {
+		p.stats.InUse().Dec()
```
Comment (asmyasnikov) [inline internal/pool/pool.go:403]: manage counters must be outside from putItem call

File: internal/pool/pool.go (hunk 3 above)
Comment (adwski) [reply internal/pool/pool.go]: Changed as you suggested and also removed `<-ctx.Done()` inside, since it doesn't do anything

File: internal/pool/defaults.go (hunk 5 above)
Comment (adwski) [reply internal/pool/defaults.go:10]: I guess it may be ok to retry instantly (createItem will wait for create timeout anyway).

File: internal/pool/pool.go (hunk 6 above)
Comment (adwski) [reply internal/pool/pool.go]: In this approach Idle() counter is showing amount of available items in queue channel, and because of this they are changed right after channel operation in getItem/putItem. Since getItem/putItem abstract queue interaction for try() and may discard item (instead of returning it (in get) or putting back to queue (in put)), managing Idle() counter outside of put/get will result in wrong value (for example when discarded items are still counted as idle).

There's not harm in moving InUse() counter outside. (Did it in last commit)

To sum up, I think It is fine if in some cases Idle+InUse < Limit, since imo counters represent stats about existing items. If item is discarded it should not be represented in stats.
